### Import Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

### Import the dataframe

In [2]:
# Import product data
path = r"C:\Users\Juliana\Documents\DATA ANALYTICS COURSE\IMMERSION\4. Python\Instacart Basket Analysis"
ords_prods_merge= pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','ords_prods_columns.pkl'))

In [34]:
# Inspect file 
ords_prods_merge.shape

(32433030, 24)

In [3]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy days        Average orders         10   
1  Regularly busy      Least busy days        Average orders         10   
2  Regularly busy      Least busy days           Most orders         10   
3      Least busy      Least busy days        Average orders         10   
4      Least busy      Least busy days           Most orders         10   

   loyalty_flag  
0  New Customer  
1  New Customer  
2  New Customer  
3  New Customer  
4  New Customer

In [5]:
# Create subset
df = ords_prods_merge[:1000000]

In [6]:
# aggregate data for the subset
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [7]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly busy days        Average orders         10   
1  Regularly busy      Least busy days        Average orders         10   
2  Regularly busy      Least busy days           Most orders         10   
3      Least busy      Least busy days        Average orders         10   
4      Least busy      Least busy days           Most orders         10   

   loyalty_flag  
0  New Customer  
1  New Customer  
2  New Customer  
3  New Customer  
4  New Customer

### 02. Aggregate data for the entire dataframe

In [8]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.215751
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170371
12               15.887622
13               16.583304
14               16.759763
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### 03. Analyze the results. How do they compare to those of the subset?

The aggregated data from the entire database naturally includes all the departments, resulting in a larger number of orders, and therefore, the mean values differ significantly. In the subset, the means range from 11.29 to 19.46, whereas in the aggregated data from the entire dataset, the range is much smaller, spanning from 15.21 to 22.9, which corresponds to department 21 (Missing).

In [9]:
df_department= pd.read_csv(os.path.join(path,'02 Data','Original Data','departments.csv'))

In [10]:
df_department.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

### 04. Create a loyalty flag for existing customers using the transform() and loc() functions.

In [11]:
# Split the data into groups based on the “user_id” column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   
10                    7.0       196.0                2.0        0.0   
11                    8.0       196.0                6.0        1.0   
12                    7.0       196.0                1.0        1.0   
13                    7.0       196.0                1.0        1.0   
14                   14.0       196.0                1.0        1.0   
15                    6.0       196.0                3.0        0.0   
16                    8.0       196.0               15.0        1.0   
17                    6.0       196.0                1.0        1.0   
18                    6.0       196.0                2.0        0.0   
19                    7.0       196.0               17.0        0.0   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda

### Deriving Columns with loc()

In [13]:
# Create flags to assign "loyalty" labels to user ID based on max order value
ords_prods_merge.loc[ords_prods_merge['max_order']> 40, 'loyalty_flag'] = 'Loyal Customer'

In [14]:
ords_prods_merge.loc[(ords_prods_merge['max_order']<= 40) & (ords_prods_merge['max_order']> 10), 'loyalty_flag'] = 'Regular Customer'

In [15]:
ords_prods_merge.loc[ords_prods_merge['max_order']<= 10, 'loyalty_flag'] = 'New Customer'

In [16]:
# Check frequencies of loyalty flags for accuracy
ords_prods_merge['loyalty_flag'].value_counts()

Regular Customer    15890458
Loyal Customer      10293434
New Customer         6249138
Name: loyalty_flag, dtype: int64

### 05. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category

In [17]:
# Group by 'prices' 
ords_prods_merge.groupby('prices')

In [18]:
ords_prods_merge['prices'].describe()

count    3.243303e+07
mean     7.794921e+00
std      4.124281e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

In [36]:
# Aggregate the product prices basic stats
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean','min','max']})

prices           
                      mean  min   max
loyalty_flag                         
Loyal Customer    7.777721  1.0  25.0
New Customer      7.804810  1.0  25.0
Regular Customer  7.802179  1.0  25.0

### 06. Create a spending flag for each user based on the average price across all their orders

In [21]:
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [22]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1  Regularly busy      Least busy days        Average orders        10   
2  Regularly busy      Least busy days           Most orders        10   
3      Least busy      Least busy days        Average orders        10   
4      Least busy      Least busy days           Most orders        10   

   loyalty_flag avg_price  
0  New Customer  6.367797  
1  New Customer  6.367797  
2  New Customer  6.367797  
3  New Customer  6.367797  
4  New Customer  6.367797  

[5 rows x 21 columns]

In [23]:
# Create flags to assign "spending" labels
ords_prods_merge.loc[ords_prods_merge['avg_price']>= 10, 'spending_flag'] = 'High Spender'

In [24]:
ords_prods_merge.loc[ords_prods_merge['avg_price']< 10, 'spending_flag'] = 'Low Spender'

In [25]:
# Check frequencies
ords_prods_merge['spending_flag'].value_counts()

Low Spender     32312346
High Spender      120684
Name: spending_flag, dtype: int64

### 07. Create and order frequency flag that marks the regularity of a user's ordering behaviour according to the median in the "days_since_prior_order" column

In [26]:
#Gorup data by user_id and create a 'media_days' column
ords_prods_merge['median_days'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [27]:
# Check grouping
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

  product_name  ...  _merge    price_range_loc     busiest_day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

          busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly busy days        Average orders        10  New Customer   
1      Least busy days        Average orders        10  New Customer   
2      Least busy days           Most orders        10  New Customer   
3      Least busy days        Average orders        10  New Customer   
4      Least busy days           Most orders        10  New Customer   

  avg_price  spending_flag median_days  
0  6.367797    Low Spender        20.5  
1  6.367797    Low Spender        20.5  
2  6.367797    Low Spender        20.5  
3  6.367797    Low Spender        20.5  
4  6.367797    Low Spender        20.5  

[5 rows x 23 columns]

In [28]:
# Create flags to assign "freqency" labels 
ords_prods_merge.loc[ords_prods_merge['median_days']> 20, 'frequency_flag'] = 'Non-frequent Customer'

In [29]:
ords_prods_merge.loc[(ords_prods_merge['median_days']<= 20) & (ords_prods_merge['median_days']> 10), 'frequency_flag'] = 'Regular Customer'

In [30]:
ords_prods_merge.loc[ords_prods_merge['median_days']<= 10, 'frequency_flag'] = 'Frequent Customer'

In [31]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN       196.0                1.0        0.0   
1                   15.0       196.0                1.0        1.0   
2                   21.0       196.0                1.0        1.0   
3                   29.0       196.0                1.0        1.0   
4                   28.0       196.0                1.0        1.0   

  product_name  ...    price_range_loc     busiest_day         busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy  Regularly busy days   
1         Soda  ...  Mid-range product  Regularly busy      Least busy days   
2         Soda  ...  Mid-range product  Regularly busy      Least busy days   
3         Soda  ...  Mid-range product      Least busy      Least busy days   
4         Soda  ...  Mid-range product      Least busy      Least busy days   

  busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0        Average orders        10  New Customer  6.367797   Low Spender   
1        Average orders        10  New Customer  6.367797   Low Spender   
2           Most orders        10  New Customer  6.367797   Low Spender   
3        Average orders        10  New Customer  6.367797   Low Spender   
4           Most orders        10  New Customer  6.367797   Low Spender   

   median_days         frequency_flag  
0         20.5  Non-frequent Customer  
1         20.5  Non-frequent Customer  
2         20.5  Non-frequent Customer  
3         20.5  Non-frequent Customer  
4         20.5  Non-frequent Customer  

[5 rows x 24 columns]

In [32]:
ords_prods_merge['frequency_flag'].value_counts()

Frequent Customer        21576963
Regular Customer          7216644
Non-frequent Customer     3639423
Name: frequency_flag, dtype: int64

### Export your dataframe as a pickle file

In [33]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_agg.pkl'))